### PR TITLE
Support creation of raw transactions with non-wallet inputs

### PR DESCRIPTION
**Pull request by dexX7**
Four new RPCs are added to create or modify raw transactions with non-wallet inputs:

- add a payload with class B encoding (`omni_createrawtx_multisig`)
- add a payload with class C encoding (`omni_createrawtx_opreturn`)
- add a transaction input (`omni_createrawtx_input`)
- add a reference output (`omni_createrawtx_reference`)
- add a change output (`omni_createrawtx_change`)

See the [JSON-RPC API documentation](https://github.com/OmniLayer/omnicore/blob/6a40683ef9bd6bdffe14772bdf2059350e79fab7/src/omnicore/doc/rpc-api.md#omni_createrawtx_opreturn).

The core components are the new TxBuilder and OmniTxBuilder classes.

See the [class documentation](https://github.com/OmniLayer/omnicore/blob/75952009b30e4d4445a953a61344c4501dad8f5b/src/omnicore/createtx.h).

Five new parsing helpers are added:

- `ParseTransaction()`
- `ParseMutableTransaction()`
- `ParsePubKeyOrAddress()`
- `ParseOutputIndex()`
- `ParsePrevTxs()`

The new classes and functions are unit tested (except RPCs).

### CODE REVIEW

**Comment by dexX7** [comment]
Re: the push a few seconds ago:

This is not an update, and there is still quite a bit to do, but now the branch is conflict free and up-to-date.

**Comment by dexX7** [comment]
I started with yet another approach, but this one I kinda like.

Previously the issue for me was that everything was stuffed into one RPC, such as class B, class C, reference, no reference, fees, reference amounts, etc. etc., which was really ugly in my opinon, and at the same time still very limiting.

The idea now is that there are several RPCs to modify raw transactions:

- add a payload with class B encoding
- add a payload with class C encoding
- add a reference output
- add a change output (and inputs)

See:

- [src/omnicore/rpcrawtx.cpp#L177-L346](https://github.com/dexX7/bitcoin/blob/wip-oc-0.10-raw-transactions/src/omnicore/rpcrawtx.cpp#L177-L346)

It's not yet finished, and I'd like to untangle it a a bit further, and see which calls are actually needed. Adding change and inputs is probably going to be seperated. The currently existing calls are already sufficient to create Omni transactions though. :)

The core components are the new `TxBuilder` and `OmniTxBuilder` classes, which are mostly finished.

See:

- [src/omnicore/createtx.h](https://github.com/dexX7/bitcoin/blob/wip-oc-0.10-raw-transactions/src/omnicore/createtx.h) (has documentation with examples)

They may be extended, and I had a lot more methods at the beginning, but I removed all methods, which were not strictly necessary. Still missing is a mechanism to add inputs via the builder, which is done via `InputsToMutableTx` on the RPC layer at the moment.

**Comment by dexX7** [comment]
@zathras-crypto: ready.

Please let me know, if there is anything to improve. :)

**Comment by dexX7** [comment]
I'd also like to add a call to modify inputs, or rather: add an input. This is something that could be used in the context of #11 (to add the payment input).

**Comment by zathras-crypto** [comment]
> @zathras-crypto: ready.

Sorry dude, missed the ping, will review again this weekend but I did play with this a little while back... Leave it with me :)

> I'd also like to add a call to modify inputs, or rather: add an input. This is something that could be used in the context of #11 (to add the payment input).

I think this is fine, but perhaps it should be restricted to Class C only where sender identification is changed to first vin only and adding inputs doesn't risk changing the sender (when using largest input by sum).

**Comment by dexX7** [comment]
> I think this is fine, but perhaps it should be restricted to Class C only where sender identification is changed to first vin only and adding inputs doesn't risk changing the sender (when using largest input by sum).

Hm.. the builder doesn't have an understanding about Omni transactions, or their senders. Transactions are built in steps, with something like `addReference()`, `addOpReturn()`, whereby each step is exposed via the API as stateless call.

It's very low level, but I'm all for better usability, though I currently don't see how. Initially I used a broader approach, where the whole transaction is created in one step, i.e. the user provides all information together, but this was awefully large and unflexible.

**Comment by dexX7** [comment]
I added `ParseOutputIndex` and `omni_createrawtx_input`.

**Comment by zathras-crypto** [comment]
Yep, that makes sense & I like the way it's done now in 'pieces' rather than a single pass, it's definitely more flexible.

I guess if the user is crafting raw transactions they should be fully aware of what they are doing and protecting them against an unintended change of sender by adding extra inputs might be a bit too much babysitting.

**Comment by dexX7** [comment]
Ideally one would build a transaction, and then validate the results with `"omni_decodetransaction"`.

File: src/omnicore/rpcrawtx.cpp [hunk 1]
```diff
@@ -69,43 +67,18 @@ Value omni_decodetransaction(const Array& params, bool fHelp)
...
+            + HelpExampleRpc("omni_decodetransaction", "\"010000000163af14ce6d477e1c793507e32a5b7696288fa89705c0d02a3f66beb3c5b8afee0100000000ffffffff02ac020000000000004751210261ea979f6a06f9dafe00fb1263ea0aca959875a7073556a088cdfadcd494b3752102a3fd0a8a067e06941e066f78d930bfc47746f097fcd3f7ab27db8ddf37168b6b52ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac00000000\", [{\"txid\":\"eeafb8c5b3be663f2ad0c00597a88f2896765b2ae30735791c7e476dce14af63\",\"vout\":1,\"scriptPubKey\":\"76a9149084c0bd89289bc025d0264f7f23148fb683d56c88ac\",\"value\":0.0001123}]")
         );
 
-    CTransaction tx;
-    if (!DecodeHexTx(tx, params[0].get_str())) {
-        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Transaction deserialization failed");
-    }
+    CTransaction tx = ParseMutableTransaction(params[0]);
```
Comment (dexX7) [inline src/omnicore/rpcrawtx.cpp]: Note to self: should be `ParseTransaction`, not `ParseMutableTransaction`.

**Comment by zathras-crypto** [comment]
> Ideally one would build a transaction, and then validate the results with "omni_decodetransaction"

Ahh of course!

**Comment by dexX7** [comment]
@zathras-crypto: did you have the chance to review?

**Comment by zathras-crypto** [comment]
> @zathras-crypto: did you have the chance to review?

I did yep and the code is great, basically what I was attempting to do was recreate existing Omni transactions via the raw interface (of course they'd be unsigned) and compare them.

I was going to loop back around and try and figure out where I went wrong as I ran into some trouble.

Here is an example of one of the transactions I built, but I'm unable to decode it via ```omni_decodetransaction``` - I haven't had a chance to dig into why as yet (looked OK from a quick look at it via ```deocderawtransaction```):

```
010000000163da7c92a0165f68fa280ccdbaaf46d5f61da7640148b40f6c9588c0161e28d10000000000ffffffff033c0f00000000000067514104e6bf7e7f3b8e1b5cbdb7ef22e893583aba36edc63871e265493fd04bd59230da6bdcbd76887863a0294f1114e02328e46b4b48b0fc364aef80f7f831b39eb1f02102c7e8649fc477c0622f2cb0fb5b3dea02a4044958b76809b8e48ecbfaec2299a452aeaa0a0000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288acaa0a0000000000001976a914fa0692278afe508514b5ffee8fe5e97732ce066988ac00000000
```

Oh, I also forgot to mention - the documentation notes the rawtx argument is optional, but it's required.  Perhaps this is just my interpretation, but I consider it a 'required' field if we have to include it (even empty).  Eg if I don't include the field because it's optional the command fails ```omni_createrawtx_opreturn "payload"```, but add the field with an empty value ```omni_createrawtx_opreturn "" "payload"``` and it's fine hence (imo) the field is required and supports empty values.  I hope that makes sense hehe :)

While on the subject, I did want to say the quality of your work is just really really professional (including commenting and documentation) - it really makes me realize that I need to "up my game" to bring my contributions up to the same standard/level of quality...

EDIT: I just did the same test transactions on another box with a fully synchronized node, I was able to decode successfully.  That makes sense now I think about it...  OK to merge :)

**Comment by dexX7** [comment]
> Here is an example of one of the transactions I built, but I'm unable to decode it [...] EDIT: I just did the same test transactions on another box with a fully synchronized node

Ah, I see. Did you explicitly provide the inputs when using `"omni_decodetransaction"`? If not, I assume this is the cause. The following should work, even on an unsynchronized box (untested):
```js
omni_decodetransaction "010000000163da7c92a0165f68fa280ccdbaaf46d5f61da7640148b40f6c9588c0161e28d10000000000ffffffff033c0f00000000000067514104e6bf7e7f3b8e1b5cbdb7ef22e893583aba36edc63871e265493fd04bd59230da6bdcbd76887863a0294f1114e02328e46b4b48b0fc364aef80f7f831b39eb1f02102c7e8649fc477c0622f2cb0fb5b3dea02a4044958b76809b8e48ecbfaec2299a452aeaa0a0000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288acaa0a0000000000001976a914fa0692278afe508514b5ffee8fe5e97732ce066988ac00000000" '[{"txid":d1281e16c088956c0fb4480164a71df6d546afbacd0c28fa685f16a0927cda63","vout":0,"scriptPubKey":"76a914ee2953140b43d86027ced62f566c46a7ed495e9e88ac","value":0.0005}]'
```

> Oh, I also forgot to mention - the documentation notes the rawtx argument is optional, but it's required.

Good you noticed this. It's one of the points I was wondering about. Currently the following approach is used:

- if there is an argument at the very end, which has a default value, and which can be omitted, then the argument is optional
- if there is an argument at the beginning or inbetween, which has a default value, and which may be set to a value, which isn't strictly matching to the underlying data type, then I call it optional, too

Let me clarify:

The call `"omni_createrawtx_opreturn"` expects `"rawtx" "payload"`, whereby `"rawtx"` refers to a transaction. An empty value (either `""` or `null`) isn't a transaction, but accepted, and by labelling it "optional" I wanted to emphasize that providing a transaction is not strictly needed.

A similar approach was used for a few other calls, for example `"omni_sendissuancecrowdsale"` (see the [docs](https://github.com/dexX7/bitcoin/blob/wip-oc-0.10-raw-transactions/src/omnicore/doc/rpc-api.md#omni_sendissuancecrowdsale)).

But I see your point and I agree, this is indeed confusing, and one may be tempted to omit optional arguments completely, which is worse than calling it "required". I'm going to update and change "optional" to "required" for those calls. :)

> While on the subject [...] it really makes me realize that I need to "up my game" to bring my contributions up to the same standard/level of quality...

Thanks, very appreciated! But no worries mate, it's a pleasure to work with you. :) While I believe we should always try to improve, content should be higher valued than style etc., which is only nice to have, but not mandatory.

As a general rule of thumb, I try to tackle it from a perspective of "what can I do to make it easier for a fellow developer, who may not be familiar with the codebase". In an ideal scenario pointing to the code or documentation should be sufficient and self-sustaining, without the need for additional explainations (and if there is need for explanations, then it may as well be documented).

**Comment by dexX7** [comment]
Updated.

**Comment by zathras-crypto** [comment]
> Did you explicitly provide the inputs when using "omni_decodetransaction"

Nope, I picked up on that afterwards - I thought "of course it doesn't work, how can we identify sender if the inputs are not in the chain?" and remembered you had provided the option to supply the inputs directly to ```omni_decodetransaction``` so no problems there :)

> A similar approach was used for a few other calls, for example "omni_sendissuancecrowdsale"

Understood, but even though things like URL or DATA may be empty, the field is still required (and noted as 'required' on the API docs).

> But I see your point and I agree, this is indeed confusing, and one may be tempted to omit optional arguments completely, which is worse than calling it "required". I'm going to update and change "optional" to "required" for those calls. :)

Thanks :)  That was my main concern, that someone would be making the call with just one parameter thinking they didn't need to include the optional parameter and not getting far - I think the change is a good call :)

> Thanks, very appreciated! But no worries mate...

You're too kind mate haha - any programming I did previously was to fill a need as part of my work in infrastructure (middleware etc) so the source never saw the light of day - I've got into some bad habits but every time I review your work I think "so polished! man my contributions have rough edges!" so if only for just taking pride in my own work I definitely think a little more effort in commenting and documenting would go a long way :)

Also I'm more than happy if you want to merge this in bud :)

**Comment by dexX7** [comment]
> (and noted as 'required' on the API docs)

Just to clarify: this was updated with this PR, and it currently still shows "optional" on the main brach.

> Also I'm more than happy if you want to merge this in bud :)

Yay, thanks! :)

**Pull request merged**